### PR TITLE
Assignment2 - Cole Stewart

### PR DESCRIPTION
**Pull request by colestewart12**

### CODE REVIEW

File: editorconfig.txt
Comment (BenjaminMichaelis) [inline editorconfig.txt:1]: This file needs to be named `.editorconfig`

Comment (colestewart12) [reply editorconfig.txt:1]: Thanks, didn't mean to commit that change

File: Logger/Logger.csproj [hunk 1]
```diff
@@ -1,8 +1,19 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (Joshua-Pedersen) [inline Logger/Logger.csproj:4]: Be sure to update to 8.0

File: Logger/Logger.csproj [hunk 2]
```diff
@@ -1,8 +1,19 @@
...
-    <LangVersion>11.0</LangVersion>
+    <LangVersion>12.0</LangVersion>
+	<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+	<EnableNETAnalyzers>True</EnableNETAnalyzers>
+	<Nullable>enable</Nullable>
+  </PropertyGroup>
```
Comment (Joshua-Pedersen) [inline Logger/Logger.csproj:9]: Don't forget to add:
```<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>```
and
    ```<TreatWarningsAsErrors>true</TreatWarningsAsErrors>```

File: Logger.Tests/Logger.Tests.csproj [hunk 3]
```diff
@@ -1,16 +1,20 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
 	<PropertyGroup>
-		<TargetFramework>net7.0</TargetFramework>
+		<TargetFramework>net8.0</TargetFramework>
 		<IsPackable>false</IsPackable>
-		<LangVersion>11.0</LangVersion>
+		<LangVersion>12.0</LangVersion>
+		<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```
Comment (Joshua-Pedersen) [inline Logger.Tests/Logger.Tests.csproj:7]: Same as in ```Logger.csproj```
Don't forget to add:
```<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>```
and
    ```<TreatWarningsAsErrors>true</TreatWarningsAsErrors>```
and
```<Nullable>enable</Nullable>```

File: Logger/LogFactory.cs [hunk 4]
```diff
@@ -2,8 +2,24 @@
...
-    public BaseLogger CreateLogger(string className)
+
+    private string? fileLoggerPath;
+
+    public void ConfigureFileLogger(string filePath)
+    {
+        fileLoggerPath = filePath;
+    }
+
+    public BaseLogger? CreateLogger()
```
Comment (Joshua-Pedersen) [inline Logger/LogFactory.cs:13]: Implementation seems fine but unsure if they would want you to keep the given signature which takes a parameter.

File: Logger/BaseLoggerMixins.cs [hunk 5]
```diff
@@ -1,6 +1,43 @@
...
+    }
+
+    public static void Debug(this BaseLogger? logger, string message, params object[] args)
+    {
+        LogWithLevel(logger, LogLevel.Debug, message, args);
+    }
+
+    public static System.Globalization.CultureInfo? CurrentCulture { get; set; }
+
+    private static void LogWithLevel(BaseLogger? logger, LogLevel logLevel, string message, params object[] args)
```
Comment (Joshua-Pedersen) [inline Logger/BaseLoggerMixins.cs:29]: Nice use of a helper method

**Review by Joshua-Pedersen — COMMENTED** [review]
## Assignment

- There is an existing `BaseLogger` class. It needs an **auto property** to hold a class name. This property should be set in the `LogFactory` using an **object initializer**. ✔
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
  Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ✔
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ❌
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌
- Ensure that you turn on EnforceCodeStyleInBuild ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ❌
- Turn on Nullability (`Nullable`) ❌
- **All of the above should be unit tested.**

## Extra Credit

- Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses `System.Console` or `System.Diagnostics.Trace` ❌
- Implement the factory pattern using static abstract methods on a Logger interface instead.

File: Logger.Tests/Logger.Tests.csproj (hunk 3 above)
Comment (colestewart12) [reply Logger.Tests/Logger.Tests.csproj:7]: Thanks for the reminder

File: Logger/Logger.csproj (hunk 1 above)
Comment (colestewart12) [reply Logger/Logger.csproj]: Thanks

**Comment by StrangeRanger** [comment]
@Joshua-Pedersen 

> Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ❌
> Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌
> Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
> Turn on Nullability (Nullable) ❌

When it comes to the analyzers/configuration-based requirements, make sure to provide a little description, such as which projects don't contain them.

Example:

> Turn on Nullability (Nullable) ❌ => Not set in the test project

**Pull request closed without merging**